### PR TITLE
feat(archive): make resource name the target path

### LINKED ISSUE #81: Small feature request: Allow naming of archive binary via pillar
### Is your feature request related to a problem?
Sort of.  Basically it's common that an upstream project will name a binary release to include a version number or platform suffix. Example: https://github.com/mikefarah/yq/releases . I want that binary named `yq` in /usr/local/bin, not `yq_linux_amd64`.

I think the target binary filename should based on the name of the pillar value, not the URL suffix.   

### Describe the solution you'd like
I'd propose the following tweaks:
```diff
--- a/packages/archives.sls
+++ b/packages/archives.sls
@@ -26,7 +26,11 @@ packages-archive-unwanted-{{ file_or_directory }}:

   # wanted 'archive' software
   {%- for package, archive in wanted_archives.items() %}
-    {%- set archivename = archive.dl.source.split('/')[-1] %}
+    {%- if archive.dl.format in ('bin',) %}
+      {%- set archivename = package %}
+    {%- else %}
+      {%- set archivename = archive.dl.source.split('/')[-1] %}
+    {%- endif %}

 packages-archive-wanted-target-{{ package }}-directory:
   file.directory:
```
```diff
--- a/pillar.example
+++ b/pillar.example
@@ -146,12 +146,14 @@ packages:
           # yamllint disable-line rule:line-length
           source: https://raw.githubusercontent.com/openstack/cinder/master/contrib/block-box/docker-compose.yml
           hashsum: 1751f8e4f6b4cddd8c4843a0f4473274
-      kubectl:
+      # If the format is "bin", the resource name will be used as the filename.
+      # Thus, the following block creates /usr/local/bin/yq
+      yq:
         dest: /usr/local/bin
         dl:
           format: bin
           # yamllint disable-line rule:line-length
-          source: https://storage.googleapis.com/kubernetes-release/release/v1.12.0/bin/darwin/amd64/kubectl
+          source: https://github.com/mikefarah/yq/releases/download/v4.13.4/yq_linux_amd64
     unwanted:
       - /usr/local/boring_archive_software
```

### Describe alternatives you've considered
n/a

### Additional context
n/a

### PR DESCRIPTION
**Pull request by mdschmitt**
### PR progress checklist (to be filled in by reviewers)

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [ ] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

Yes, though it's small.  If the archive package is of type `bin`, the name of the downloaded object will be changed if the resource name in the pillar is not the same as the last part of the URL.

### Related issues and/or pull requests

Fixes #81 

### Describe the changes you're proposing

<see description of #81 >

### Pillar / config required to test the proposed changes

[example pillar](pillar.example) details the change, the change itself is in [archives.sls](packages/archives.sls)

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [x] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

**Comment by mdschmitt** [comment]
@myii , you've got the last few commits. Maybe you can tell me why these checks are failing?

**Comment by myii** [comment]
> @myii , you've got the last few commits. Maybe you can tell me why these checks are failing?

@mdschmitt Yes, those are unrelated to this PR and nothing recent that was changed.  It's failing to install `dxpy` using `pip`, so we'll probably just switch to some other package for testing purposes.  @javierbertoli IIRC, you were planning to adjust this?

**Comment by mdschmitt** [comment]
@javierbertoli , no pressure on this, just want to make sure it stays on the radar.  Is there anything I can do to help?

**Comment by mdschmitt** [comment]
@javierbertoli , I know it's been a little bit, but can I do anything to help fix this CI and get this relatively small change merged in?

File: packages/archives.sls [hunk 1]
```diff
@@ -26,7 +26,11 @@ packages-archive-unwanted-{{ file_or_directory }}:
 
   # wanted 'archive' software
   {%- for package, archive in wanted_archives.items() %}
-    {%- set archivename = archive.dl.source.split('/')[-1] %}
```
Comment (javierbertoli) [inline packages/archives.sls:29]: I would suggest a change here, to make the destination name configurable through a variable (like `name`, `destname` or `binname`?) that will also make it backward compatible:

```sls
{%- set archivename = archive.dl.source.split('/')[-1] if 'destname' not in archive else archive.destname %}
```

This way, we can define an arbitrary destination name like

```yaml
packages:
  archives:
    wanted:
      yq:
        dest: /usr/local/bin
        destname: my_yq
          dl:
            format: bin
            source: https://github.com/mikefarah/yq/releases/download/v4.13.4/yq_linux_amd64
```

Another backward-compatible alternative is to add a boolean parameter like `destname_as_package: true` (or any better var name you can think of) and then

```sls
{%- set use_packagename_as_binaryname = archive.destname_as_package | default(false) %}
{%- set archivename = package if use_packagename_as_binaryname else archive.dl.source.split('/')[-1] %}
```
```yaml
packages:
  archives:
    wanted:
      yq:
        dest: /usr/local/bin
        archive.destname_as_package: true
          dl:
            format: bin
            source: https://github.com/mikefarah/yq/releases/download/v4.13.4/yq_linux_amd64
```
Although I think this 2nd option is less flexible that the 1st one.

wdyt?

**Review by javierbertoli — CHANGES_REQUESTED** [review]
@mdschmitt I really sorry that I didn't come back to this issue to get it added.

Check the change I suggest and let me know what do you think of them?

File: packages/archives.sls (hunk 1 above)
Comment (mdschmitt) [reply packages/archives.sls:29]: I think you're right and that's a good way to keep everyone happy.  Committed this, CI should be running soon.

Comment (javierbertoli) [reply packages/archives.sls:29]: @mdschmitt ?

**Review by javierbertoli — APPROVED** [review]
It lgtm

**Comment by javierbertoli** [comment]
Thanks for this, @mdschmitt !

**Pull request merged**